src/command_executor.cc

```diff
@@ -75,30 +87,40 @@ void CommandExecutor::WaitChildExit(pid_t pid) {
 }
 
 int CommandExecutor::Execute(const std::string &line) {
-  CommandParseResult command_parse_result =
+  std::vector<CommandParseResult> command_parse_result_list =
       parser_.ParseUserInputLine(line);
-  if (build_in_.Exist(command_parse_result.command)) {
-    return build_in_.Execute(command_parse_result.command,
-                             command_parse_result.args);
-  }
+  bool is_pipe_redirect = command_parse_result_list.size() != 1;
+  int pipe_fds[2];
+  pipe(pipe_fds);
+  for (int i = 0; i < command_parse_result_list.size(); i++) {
+    CommandParseResult command_parse_result = command_parse_result_list[i];
+    if (build_in_.Exist(command_parse_result.command)) {
+      return build_in_.Execute(command_parse_result.command,
+                               command_parse_result.args);
+    }
+    pid_t pid = fork();
+    if (pid == ERROR_CODE_SYSTEM) {
+      // error
+      utils::PrintSystemError(std::cerr);
+      return ERROR_CODE_DEFAULT;
+    }
 
-  pid_t pid = fork();
-  if (pid == ERROR_CODE_SYSTEM) {
-    // error
-    utils::PrintSystemError(std::cerr);
-    return ERROR_CODE_DEFAULT;
-  }
-
-  if (pid == 0) {
-    return ProcessChild(command_parse_result);
-  } else {
-    WaitChildExit(pid);
+    auto cmd_number = i;
+    if (pid == 0) {
+      return ProcessChild(command_parse_result, is_pipe_redirect, pipe_fds,
+                          cmd_number);
+    } else {
+      close(pipe_fds[1]);
+      WaitChildExit(pid);
```

Comment:
Do you want to wait both these two command to exit or can we just wait for last command, because these two command should be chain as a pipeline, so only wait for last command should be enough.